rfcs/0013-avalanche.md

```diff
@@ -0,0 +1,98 @@
+to address some of them.  
+
+* Most importantly, there needs to be a mechanism to figure out which nodes belong in what layer of this avalanche. We need a Stake weighted selection for the nodes. This is missing and is very important. 
+* Leaders should only broadcast to layer 1 nodes and the network should be split (ClusterInfo can provide leader’s layer-1 tvu peers) 
+Layer 1 nodes should broadcast to all peers at the same layer and to a subset of peers at lower layers. 
+(ClusterInfo can provide Layer 1’s peers and layer 2 sections)
+* Layer 2 nodes should only retransmit amongst themselves. They should also be sectioned off and only need to communicate with each other instead of all layer 2 nodes. 
+(ClusterInfo can provide which nodes are in this node’s neighborhood)
+* 200 fanout from leader to layer-1 and layer-1 to layer-2.
+* Need to make sure layer-2 has a neighborhood (stake based) and that there can be some overlap. 
+
+
+### The proposed changes
+The idea is to fanout to 200 nodes from leader to layer 1 and then from layer 1 to layer 2. Layer 2 nodes should be semi-aware of this fanout so that they only talk to nodes in their "neighborhood". 
+This will result in the formation of multiple layer-2 neighborhoods that only care about communicating with each other and then the leader to send their votes. 
+Each layer 1 node only needs to broadcast its data to a single node in each neighborhood. With this in mind, the following changes will be needed to the two main stages.
+
+#### A Stake weighted selection mechanism:
+Need to come up with some logic to filter a set of nodes based on a stake range or number boundary. For the leader to find its layer 1 nodes, we need the biggest stake holders. 
+For each layer 1 node to find its layer 1 peers, the same logic *should* work but each layer 1 node also needs to find its layer 2 group. This needs some thought.
+Stake weights should work once we know the weight of the last layer 1 node. Layer 2 nodes can be "ordered" by weight and all layer 1 nodes should be able to avoid stepping over each others layer 2 nodes.    
+ 
+#### Broadcast stage:
+Broadcast Stage might need a bank to figure out stakes and can hand this off to ClusterInfo to figure out the top 200/85% stake holders. 
+These top stake holders will be considered Layer 1. For the leader this step should be pretty straightforward and can be achieved with a `get_top_tvus()` call to sort the top stake holders' tvus. 
+
+#### Replicate stage:  
+The biggest challenge is to update replicate stage and make it "layer aware"; i.e using the bank each layer 1 node can figure out which layer 2 nodes to send blobs to with minimal overlap. 
+The plan to avoid this overlap
+will be to broadcast using indices based on `((node.layer_1_index) % (layer-2_neighborhood_size) * cur_neighborhood_index)` where `cur_neighborhood_index` is loop index in `num_neighborhoods`.
```

Comment:
In my opinion, this is the most important calculation to grok perfectly. I think you should move the block diagram down here, and greatly expand this section. I think there's more match that needs to be worked out here and that you might kick off this project by writing a simulator to ensure you have this perfect. The diagram shows how it works with 3 layers and a multiple of 2. It needs to work with `N` layers and a multiple of `M`.  There could be one pair of diagrams for when `N` is 2 and `M` goes from 2 to 3. And then another diagram for when `M` is pinned, and `N` increases by one.  That is, there's a base case and inductive case for each variable. We need to be sure that data isn't lost when, for example, the "neighborhood size" isn't a multiple of `M`.

Reply:
+1 for the simulator, I think that'll help test my assumptions. I'll look at how @aeyakovenko did the gossip simulation and try to use something like that. In theory this should not miss any nodes at all. 
But if the assumption is to work with `N` layers then I might have to rethink this. I was planning for 2 layers (should be able to cover 40k nodes with a fanout of 200). I could leave that configurable in the code but then we might run into nodes running different versions (potentially) with different values. Maybe gossip can spread the fanout and neighborhood configuration around?

Reply:
Yes, that'd be great if the fanout multiple were configurable via an on-chain program.  Dumping config data in gossip is a dirty hack and will probably just have you chasing superficial bugs as different nodes have a different snapshot of the data in gossip. By putting it in a program, you can gossip just the entry height at which all nodes should update their configurations. Squeaky clean.